src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +400,122 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
+                var newCompletedItems = Math.Max(0, completedItems - progressTracker.CompletedItems);
 
-                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                    return document.Project.Solution.Workspace.TryApplyChanges(newDoc.Project.Solution, progressTracker);
+                for (var i = 0; i < newCompletedItems; i++)
+                {
+                    progressTracker.ItemCompleted();
+                }
+            });
+
+            // Prepopulate the solution progress tracker with the total number of documents to process
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = solution.GetProject(projectId);
+                if (project.LanguageServices.GetService<ICodeCleanupService>() == null)
+                {
+                    continue;
                 }
+
+                progressTracker.AddItems(project.DocumentIds.Count);
+            }
+
+            foreach (var projectId in solution.ProjectIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var project = solution.GetProject(projectId);
+                var newProject = await FixProjectAsync(project, enabledFixIds, projectProgressTracker, cancellationToken).ConfigureAwait(false);
+                solution = newProject.Solution;
             }
+
+            return solution;
+        }
+
+        private async Task<Project> FixProjectAsync(Project project, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (project.LanguageServices.GetService<ICodeCleanupService>() == null)
+            {
+                return project;
+            }
+
+            progressTracker.AddItems(project.DocumentIds.Count);
+            foreach (var documentId in project.DocumentIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var document = project.GetDocument(documentId);
+                progressTracker.Description = document.Name;
+
+                // FixDocumentAsync reports progress within a document, but we limit progress reporting for a project
+                // to the current document.
+                var documentProgressTracker = new ProgressTracker();
+
+                var fixedDocument = await FixDocumentAsync(document, enabledFixIds, documentProgressTracker, cancellationToken).ConfigureAwait(false);
+                project = fixedDocument.Project;
+                progressTracker.ItemCompleted();
+            }
+
+            return project;
+        }
+
+        private async Task<Document> FixDocumentAsync(Document document, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (document.IsGeneratedCode(cancellationToken))
```

Comment:
I see this generated code kind of checks for fixer in arbitary places.. we probably need to discuss and create some kind of central thing? otherwise, this behavior seems hard to make consistent?

Reply:
I'll make sure a follow-up issue is filed for this one before resolving it.

Reply:
➡️ #34418